File: Dynamic-Programming/279-M-Perfect-Squares/perfect_squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        OPT = [n] * (n+1)
        OPT[0] = 0

        for target in range(1, n+1):
            for s in range(1, target+1):
                square = s*s
                if target - square < 0:
                    break
                OPT[target] = min(OPT[target], OPT[target-square] + 1)
        return OPT[n]

File: Dynamic-Programming/279-M-Perfect-Squares/perfect_squares.c
class Solution {
public:
    int numSquares(int n) {
        vector<int> OPT(n+1, n);
        OPT[0] = 0;

        for (int i=1; i<n+1; i++) {
            for (int j=1; j<i+1; j++) {
                if (i-(j*j) < 0)
                    break;
                OPT[i] = min(OPT[i], 1+OPT[i-(j*j)]);
            }
        }
        return OPT[n];
    }
};